Cooler cload

jianhong opened the pull request:
## PR checklist

Closes #518
I massed up the files when merge the upstream conflics.

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `<SOFTWARE>.version.txt` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

grst commented:
No tests were ran, I think the PR misses an entry in `pytest_modules.yml`. I added one in one of the other cooler PRs, so you can have a look there how it needs to look like.

jianhong commented:
> No tests were ran, I think the PR misses an entry in `pytest_modules.yml`. I added one in one of the other cooler PRs, so you can have a look there how it needs to look like.

I modified pytest_modules.yml.

grst commented:
Same as in the other PR, please update to `versions.yml`

sguizard commented on tests/modules/cooler/cload/test.yml:
```diff
@@ -0,0 +1,9 @@
+- name: cooler cload test_cooler_cload
+  command: nextflow run tests/modules/cooler/cload -entry test_cooler_cload -c tests/config/nextflow.config
+  tags:
+    - cooler
+    - cooler/cload
+  files:
+    - path: output/cooler/test_pairix.2000000.cool
+    - path: output/cooler/test_pairs.2000000.cool
+    - path: output/cooler/test_tabix.2000000.cool
```
Would be modified if the code suggestions are applied.

sguizard commented on tests/modules/cooler/cload/main.nf:
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { COOLER_CLOAD } from '../../../../modules/cooler/cload/main.nf' addParams( options: [:] )
+include { COOLER_CLOAD
+ as COOLER_CLOAD_PAIRS } from '../../../../modules/cooler/cload/main.nf' addParams( options: [args:'pairs --chrom1 1 --pos1 2 --chrom2 4 --pos2 5 -N'] )
+include { COOLER_CLOAD
+ as COOLER_CLOAD_TABIX } from '../../../../modules/cooler/cload/main.nf' addParams( options: [args:'tabix'] )
+
+
+workflow test_cooler_cload {
+
+    input = [ [ id:'test_pairix', single_end:false ], // meta map
+             file(params.test_data['generic']['cooler']['test_pairix_pair_gz'], checkIfExists: true),
+             file(params.test_data['generic']['cooler']['test_pairix_pair_gz_px2'], checkIfExists: true)]
+    sizes = file(params.test_data['generic']['cooler']['hg19_chrom_sizes'], checkIfExists: true)
+    bin_size = 2000000
+    COOLER_CLOAD ( input, bin_size, sizes )
+
+    input3 = [ [ id:'test_pairs', single_end:false ], // meta map
+             file(params.test_data['generic']['cooler']['test_pairs_pair'], checkIfExists: true),
+             []]
+    COOLER_CLOAD_PAIRS ( input3, bin_size, sizes )
+
+    input4 = [ [ id:'test_tabix', single_end:false ], // meta map
+             file(params.test_data['generic']['cooler']['test_tabix_pair_gz'], checkIfExists: true),
+             file(params.test_data['generic']['cooler']['test_tabix_pair_gz_tbi'], checkIfExists: true)]
+    COOLER_CLOAD_TABIX ( input4, bin_size, sizes )
+}
```
Added pairix tool for the first test and separate each test into a workflow
```suggestion
#!/usr/bin/env nextflow

nextflow.enable.dsl = 2

include { COOLER_CLOAD }                       from '../../../../modules/cooler/cload/main.nf' addParams( options: [args:'pairix'] )
include { COOLER_CLOAD as COOLER_CLOAD_PAIRS } from '../../../../modules/cooler/cload/main.nf' addParams( options: [args:'pairs --chrom1 1 --pos1 2 --chrom2 4 --pos2 5 -N'] )
include { COOLER_CLOAD as COOLER_CLOAD_TABIX } from '../../../../modules/cooler/cload/main.nf' addParams( options: [args:'tabix'] )


workflow test_cooler_cload_pairix {

    input = [ [ id:'test_pairix', single_end:false ], // meta map
             file(params.test_data['generic']['cooler']['test_pairix_pair_gz'], checkIfExists: true),
             file(params.test_data['generic']['cooler']['test_pairix_pair_gz_px2'], checkIfExists: true)]

    sizes    = file(params.test_data['generic']['cooler']['hg19_chrom_sizes'], checkIfExists: true)
    bin_size = 2000000

    COOLER_CLOAD ( input, bin_size, sizes )

}

workflow test_cooler_cload_pairs {

    input = [ [ id:'test_pairs', single_end:false ], // meta map
             file(params.test_data['generic']['cooler']['test_pairs_pair'], checkIfExists: true),
             []]

    sizes    = file(params.test_data['generic']['cooler']['hg19_chrom_sizes'], checkIfExists: true)
    bin_size = 2000000

    COOLER_CLOAD_PAIRS ( input, bin_size, sizes )

}

workflow test_cooler_cload_tabix {

    input = [ [ id:'test_tabix', single_end:false ], // meta map
             file(params.test_data['generic']['cooler']['test_tabix_pair_gz'], checkIfExists: true),
             file(params.test_data['generic']['cooler']['test_tabix_pair_gz_tbi'], checkIfExists: true)]

    sizes    = file(params.test_data['generic']['cooler']['hg19_chrom_sizes'], checkIfExists: true)
    bin_size = 2000000

    COOLER_CLOAD_TABIX ( input, bin_size, sizes )

}

}
```

sguizard commented on tests/config/test_data.config:
```diff
@@ -255,6 +255,14 @@ params {
+            'cooler'{
+                test_pairix_pair_gz                           = "${test_data_dir}/genomics/homo_sapiens/cooler/cload/hg19/hg19.GM12878-MboI.pairs.subsample.blksrt.txt.gz"
+                test_pairix_pair_gz_px2                       = "${test_data_dir}/genomics/homo_sapiens/cooler/cload/hg19/hg19.GM12878-MboI.pairs.subsample.blksrt.txt.gz.px2"
+                test_pairs_pair                               = "${test_data_dir}/genomics/homo_sapiens/cooler/cload/hg19/hg19.sample1.pairs"
+                test_tabix_pair_gz                           = "${test_data_dir}/genomics/homo_sapiens/cooler/cload/hg19/hg19.GM12878-MboI.pairs.subsample.sorted.possrt.txt.gz"
+                test_tabix_pair_gz_tbi                       = "${test_data_dir}/genomics/homo_sapiens/cooler/cload/hg19/hg19.GM12878-MboI.pairs.subsample.sorted.possrt.txt.gz.tbi"
+                hg19_chrom_sizes                              = "${test_data_dir}/genomics/homo_sapiens/cooler/cload/hg19/hg19.chrom.sizes"
```
```suggestion
                test_pairs_pair                               = "${test_data_dir}/genomics/homo_sapiens/cooler/cload/hg19/hg19.sample1.pairs"
                test_tabix_pair_gz                            = "${test_data_dir}/genomics/homo_sapiens/cooler/cload/hg19/hg19.GM12878-MboI.pairs.subsample.sorted.possrt.txt.gz"
                test_tabix_pair_gz_tbi                        = "${test_data_dir}/genomics/homo_sapiens/cooler/cload/hg19/hg19.GM12878-MboI.pairs.subsample.sorted.possrt.txt.gz.tbi"
                hg19_chrom_sizes                              = "${test_data_dir}/genomics/homo_sapiens/cooler/cload/hg19/hg19.chrom.sizes"
```

sguizard commented on modules/cooler/cload/main.nf:
```diff
@@ -0,0 +1,55 @@
+    path "versions.yml"                           , emit: versions
+
+    script:
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    def args     = options.args.tokenize()
+    def tool     = (options.args.contains('hiclib')) ? "hiclib" :
+        (options.args.contains('tabix')) ? 'tabix' :
+        (options.args.contains('pairs')) ? 'pairs' : 'pairix'
+    def nproc    = tool == "pairix" || tool == 'tabix'? "--nproc ${task.cpus}" : ''
+    args.removeIf { it.contains('hiclib') }
+    args.removeIf { it.contains('tabix') }
+    args.removeIf { it.contains('pairs') }
+    args.removeIf { it.contains('pairix') }
+
+    """
+    cooler cload $tool \\
+        ${args.join(' ')} \\
```
I think the code could be simpler if the tool is passed with `$option.args`
```suggestion
    def prefix = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
    def nproc  = options.args.contains('pairix') || options.args.contains('tabix')? "--nproc ${task.cpus}" : ''

    """
    cooler cload \\
        $options.args \\
```

sguizard reviewed (changes requested):
I think the code could be simplified. 
I added code suggestion in this way. 
I tested the modifications locally and seems to works. You might check if those does not impact the results. 
The three tests must be separated into different workflows.
